### ISSUE TITLE
VerifyError with CompletionStage parameters

### ISSUE THREAD

**Issue by HaloFour** [electronicarts/ea-async#1]
First I'd like to congratulate you on going 1.0.  I've been watching this project since it was a part of Orbit last year with anticipation as I currently write a lot of asynchronous code in Java and having come from C# found it incredibly tedious, even after the addition of `CompletableFuture<T>`.

I am currently trying to play with a sample project to get my feet wet and I'm getting a `VerifyError` when I try to run the project.  I have the two following source files:

##### Main.java
```java
package com.halofour;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class Main {
    public static void main(String[] args) throws Throwable {
        CompletableFuture<String> future = new CompletableFuture<>();
        CompletionStage<String> stage = MyAsyncClass.testAsync(future);
    }
}
```

##### MyAsyncClass.java
```java
package com.halofour;

import java.util.concurrent.CompletionStage;

import static com.ea.async.Async.await;
import static java.util.concurrent.CompletableFuture.completedFuture;

public class MyAsyncClass {
    public static CompletionStage<String> testAsync(CompletionStage<String> stage) {
        String result = await(stage);
        return completedFuture(result);
    }
}
```

And my pom file is as follows:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <name>sandbox</name>
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.halofour</groupId>
    <artifactId>sandbox</artifactId>
    <version>1.0-SNAPSHOT</version>

    <build>
        <resources>
            <resource>
                <directory>src/main/resources</directory>
                <filtering>true</filtering>
                <includes>
                    <include>**</include>
                </includes>
            </resource>
        </resources>

        <plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>buildnumber-maven-plugin</artifactId>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.18.1</version>
                <configuration>
                    <forkCount>1</forkCount>
                    <systemPropertyVariables>
                        <unit-test>true</unit-test>
                    </systemPropertyVariables>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.3</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                    <maxmem>512m</maxmem>
                    <meminitial>128m</meminitial>
                </configuration>
            </plugin>
            <plugin>
                <artifactId>maven-assembly-plugin</artifactId>
                <version>2.5.5</version>
                <configuration>
                    <archive>
                        <manifest>
                            <mainClass>com.halofour.Main</mainClass>
                            <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
                        </manifest>
                        <manifestEntries>
                            <Implementation-Build>${project.version}</Implementation-Build>
                            <Build-Version>${project.version}</Build-Version>
                            <Build-Time>${maven.build.timestamp}</Build-Time>
                        </manifestEntries>
                    </archive>
                    <descriptorRefs>
                        <descriptorRef>jar-with-dependencies</descriptorRef>
                    </descriptorRefs>
                    <finalName>sandbox</finalName>
                </configuration>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>single</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>com.ea.async</groupId>
                <artifactId>ea-async-maven-plugin</artifactId>
                <version>1.0.0</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>instrument</goal>
                            <goal>instrument-test</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

    <dependencies>
        <dependency>
            <groupId>com.ea.async</groupId>
            <artifactId>ea-async</artifactId>
            <version>1.0.0</version>
            <scope>compile</scope>
        </dependency>
    </dependencies>
</project>
```

As you can see there is not a lot going on here.  However, when I build the project and run it I get the following error:

```
Exception in thread "main" java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    com/halofour/MyAsyncClass.async$testAsync(Ljava/util/concurrent/CompletionStage;Ljava/util/concurrent/CompletionStage;ILjava/lang/Object;)Ljava/util/concurrent/CompletableFuture; @42: invokevirtual
  Reason:
    Type 'java/util/concurrent/CompletionStage' (current frame, stack[0]) is not assignable to 'java/util/concurrent/CompletableFuture'
  Current Frame:
    bci: @42
    flags: { }
    locals: { 'java/util/concurrent/CompletionStage', 'java/util/concurrent/CompletionStage', 'java/util/concurrent/CompletionStage', 'java/lang/Object' }
    stack: { 'java/util/concurrent/CompletionStage', 'java/util/function/Function' }
  Bytecode:
    0x0000000: 1caa 0000 0000 004f 0000 0000 0000 0001
    0x0000010: 0000 0017 0000 004b 2a59 b900 1301 00b6
    0x0000020: 0019 9a00 194d 2cb8 001f b600 232a 2c11
    0x0000030: 0001 ba00 3700 00b6 003a b0b9 0013 0100
    0x0000040: b600 3ec0 0040 4c2b b800 43b0 2ba7 ffee
    0x0000050: bb00 4a59 b700 4bbf                    
  Stackmap Table:
    full_frame(@24,{Object[#15],Object[#15],Integer,Object[#4]},{})
    full_frame(@59,{Object[#15]},{Object[#15]})
    full_frame(@76,{Object[#15],Object[#15],Integer,Object[#4]},{})
    full_frame(@80,{Object[#15],Object[#15],Integer,Object[#4]},{})

	at com.halofour.Main.main(Main.java:13)
```

If I change the parameter type of the first parameter in the `MyAsyncClass#testAsync` method from `CompletionStage<T>` to `CompletableFuture<T>` it does seem to work correctly.

**Comment by JoeHegarty** [comment]
Looks like this is just an edge case for the code generation that we haven't ran into before.
I'll let @DanielSperry take a look as he's the resident expert on async.

Thanks!
Joe

**Comment by DanielSperry** [comment]
I'm on it.

**Comment by DanielSperry** [comment]
@HaloFour, Thanks for the report and for the testcase.
I fixed the problem with CompletionStage arguments. 
And incorporated your pull request with the test case.

**Comment by JoeHegarty** [comment]
This fix is in 1.0.1 and has been pushed to maven central. 
Thanks for reporting it!
